utils/getValidFilename.js

```diff
@@ -0,0 +1,37 @@
+function getUnreservedFilename(filename) {
+    // list of reserved filenames on different operating systems:
+    // https://stackoverflow.com/a/31976060
+    //
+
+    const reservedFilenames = ["CON", "PRN", "AUX", "NUL",
+        "COM1", "COM2", "COM3", "COM4", "COM5", "COM6", "COM7","COM8", "COM9",
+        "LPT1", "LPT2", "LPT3", "LPT4", "LPT5", "LPT6", "LPT7", "LPT8", "LPT9"];
+    
+    for (const reservedFilename of reservedFilenames) {
+        if (filename === reservedFilename) return (`${filename}-unreserved`);
+    }
+
+    return filename;
+}
+
+function getValidFilename(filename) {
+    let validFilename = getUnreservedFilename(filename);
+
+    const singleFiguresToString = ["0", "1", "2", "3", "4", "5", "6", "7", "8", "9"];
+
+    // generated with https://stackoverflow.com/a/24597663
+    const upperCaseCharacters = ["A", "B", "C", "D", "E", "F", "G", "H", "I", "J", "K", "L", "M", "N", "O", "P", "Q", "R", "S", "T", "U", "V", "W", "X", "Y", "Z"];
+    const lowerCaseCharacters = upperCaseCharacters.map( (e) => e.toLowerCase() );
+
+    const allowedCharacters = singleFiguresToString.concat(upperCaseCharacters).concat(lowerCaseCharacters);
+
+    for (const char of validFilename) {
+        if (!allowedCharacters.includes(char)) {
+            validFilename = validFilename.split(char).join("-");
+        }
+    }
```

Comment:
This whole function is an awe full complicated way to write

```
return getUnreservedFilename(filename).replaceAll(/[^\w]/g, "-")
```

Also i dont like how this function is both calling an abstract `getUnreservedFilename` but also does a tons of string replaces. I would rather see
```
getValidFilename(filename) {
  const validFilename = getUnreservedFilename(filename);
  const charSafeFilename = clearInvalidChars(filename);
  return charSafeFilename;
}
```

Reply:
yes the abstraction was not so good, i wanted to make a 3rd function at first but i thought it was too much, but you're right it's better this way

also i would love to learn regexp, this is so awesome

except that i didnt find the replaceAll method, i read that if a regexp is supplied as input, the replace is global, but if a string is supplied as input in .replace then only the first occurence of the character is replaced, that's why i chose to use .split(char).join("-")
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/replace

Reply:
i tested it and it works!

<img width="157" alt="regexp replace" src="https://user-images.githubusercontent.com/38690718/87905977-05bb7b00-ca62-11ea-8cd3-79e04a511209.PNG">

Reply:
oh actually i tested more in details, and all works even than my code

<img width="403" alt="regexp replace v2 works for all except underscore" src="https://user-images.githubusercontent.com/38690718/87908221-57fe9b00-ca66-11ea-93c5-8302d0d69cc5.PNG">

the 1st source i linked especially mentionned whitelisted characters to include:

>Letters (a-z A-Z) - Unicode characters as well, if needed
>Digits (0-9)
>Underscore (_)
>Hyphen (-)
>Space
>Dot (.)

https://stackoverflow.com/a/29676339

Reply:
also yargs understands dots as objects and their properties, so better also replaced it with a hyphen too:
https://github.com/yargs/yargs/blob/master/docs/tricks.md#objects

Reply:
anyways after this PR is merged i'll learn regexp, wanted to do it for a long while now :)

Reply:
also actually about abstraction there is a nuance, reserved filename is an exact string match of a specific name (ex: COM1), so fixed reserved filename (ex: COM1-unreserved) does not need to be tested for invalid characters as ouput is known not to contain any

so it looks like 2 condtional branches actually, i'll refactor that :)

Reply:
But if you pump the strings trough both methods it'll be valid right? Why need to branch here?

Reply:
If anything i would remove the reserved filename thing, that seems like a user input error, the user should know what filenames his filesystem supports.

Reply:
Re: Yargs, the string has already been parsed by yargs at this point, i dont think we need to care about that anymore.

Reply:
Also yargs sees the flag as object, not the value.

Reply:
Another thing: Yargs can convert the input to number, but that is a valid filename, so the check for string is not correct.

Reply:
And since i'm writing a long list: Yargs seems to have really nice array support, we could add that for other options like komi's and boarsizes etc.

Reply:
yes yargs is very nice, we should definitely make use of some of its nice features like array and object support (also boolean --no-option flag for example)

Reply:
yes reseved filename is actually probably never going to be useful, and is system specific, so now the getValidFilename function is so small (2 lines) and also not used anywhere else, so i moved it it config.js instead of its own module

Reply:
so i added a .string in yargs's getArgv.js to make sure our input is string (number-only will be parsed as a string, and empty will not be a boolean true but should be an empty string (did not test it yet)

i'll import these changes in #332 and update unit testing for these

Reply:
also unrelated but yesterday i discovered git command-line in vs code is better than ubuntu shell command-line, with vs code's command line git remembers password and username, and most likely has other features too

![vs code git command line](https://user-images.githubusercontent.com/38690718/87917917-f8a88700-ca75-11ea-8fdc-86ae0bed382f.png)